File: leetcode/binary-search/0034_find_first_and_last.py
# LC 34. Find First and Last Position of Element in Sorted Array | Medium
class Solution:
    def searchRange(self, nums, target):
        def find(leftBias):
            l, r, idx = 0, len(nums) - 1, -1
            while l <= r:
                m = (l + r) // 2
                if nums[m] == target:
                    idx = m
                    if leftBias: r = m - 1
                    else: l = m + 1
                elif nums[m] < target: l = m + 1
                else: r = m - 1
            return idx
        return [find(True), find(False)]

File: leetcode/binary-search/0034_find_first_and_last.cpp
// LC 34. Find First and Last Position | Medium
#include <vector>
using namespace std;

class Solution {
    int find(vector<int>& nums, int target, bool leftBias) {
        int l = 0, r = nums.size() - 1, idx = -1;
        while (l <= r) {
            int m = l + (r - l) / 2;
            if (nums[m] == target) { idx = m; if (leftBias) r = m - 1; else l = m + 1; }
            else if (nums[m] < target) l = m + 1; else r = m - 1;
        }
        return idx;
    }
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        return {find(nums, target, true), find(nums, target, false)};
    }
};
